### PR TITLE
Feature/benchmark

### PR DESCRIPTION
**Pull request by hroest**
- add benchmarking tools
- improve memory calculation on Linux
- add cached mzML to FileConverter

this should help in benchmarking our improvements and benchmark against other projects

### CODE REVIEW

**Comment by hroest** [comment]
benchmarking can be done with 

```
FileConverter -in test.mzML -write_mzML_index -out test.idx.mzML
FileConverter -in test.mzML -out test.c.mzML.cachedMzML

TICCalculator -in test.c.mzML.cachedMzML -read_method cached -threads 4
TICCalculator -in test.idx.mzML -read_method indexed -threads 4
TICCalculator -in test.mzML -read_method regular -threads 4 
TICCalculator -in test.mzML -read_method streaming -threads 4 
```

File: src/utils/TICCalculator.cpp [hunk 1]
```diff
@@ -0,0 +1,310 @@
+// --------------------------------------------------------------------------
+//                   OpenMS -- Open-Source Mass Spectrometry
+// --------------------------------------------------------------------------
+// Copyright The OpenMS Team -- Eberhard Karls University Tuebingen,
+// ETH Zurich, and Freie Universitaet Berlin 2002-2013.
```
Comment (cbielow) [inline src/utils/TICCalculator.cpp]: 2016

File: src/topp/FileConverter.cpp [hunk 2]
```diff
@@ -452,6 +526,27 @@ class TOPPFileConverter :
       if (fm.size() > 0) EDTAFile().store(out, fm);
       else if (cm.size() > 0) EDTAFile().store(out, cm);
     }
+    else if (out_type == FileTypes::CACHEDMZML)
+    {
+      // Determine output path for meta information (empty mzML)
+      std::vector<String> split_out;
```
Comment (cbielow) [inline src/topp/FileConverter.cpp]: there is a considerable bit of code duplication when handling the paired files.
Maybe you can find a way to put that into a function?

**Comment by cbielow** [comment]
good stuff!
What are the numbers currently like?! Is there a result somewhere?

+1 if minor fixes above are adressed

**Comment by hroest** [comment]
@cbielow it is hard to report absolute numbers as it will depend on your system, but this code was used in the PLOS One paper:

http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0125108

so in Release 1.11, we could process ca 1 M peaks/second. Now using the improved OpenMS, this increases to about 4.5-5 M peaks/second, so a 4-fold improvement in speed and removal of `O(n)` memory performance down to `O(1)`. So we can use less memory and are faster, but quite similar to other software such as proteowizard.

Using multiple threads, this increases to about 7-8 M peaks/second for the regular implementation and up to 40 M peaks/second for indexed mzML. So especially indexedmzML is quite fast when processed in parallel as all actions including reading, decoding and processing can be done completely in parallel. Here the benchmark is to compute the TIC, so depending on where the bottleneck is, this may of course change. 

The caching option is way beyond that, it can process 250-300 M peaks/second so its another factor 10 faster than all the other approaches but it requires a different file format. At this point, file access is just a no-brainer and everything else will be slow compared to file access. Even SWATH files with about 4 billion peaks in them (60 GB file) are processed in 15 seconds.

**Comment by hroest** [comment]
ok, I added some more code (I forgot the parallel code somehow ... ) and addressed the comments above.

**Comment by hroest** [comment]
tests and style seem to work, should we add it? 

@cbielow if you are interested I can send you the excel sheet I used to generate the figures in the PLOS One paper and we could start comparing OpenMS releases from now on by running the TICCalculator from the release build on the same file. Counting TIC is not the best benchmark but its a start

**Comment by hroest** [comment]
@jpfeuffer any thoughts on whether we could add this to nightly tests or do you think this needs to be done manually as the server load could influence the results?

**Comment by jpfeuffer** [comment]
I can not guarantee that we can keep the server load always the same for these benchmarks but usually we try that different builds do not overlap anyway. Sometimes there might be low-footprint user interactions, too.
I also wondered if it makes sense that we create a new folder in /tests/ called benchmark, such that it can easily be en/disabled via some cmake variable like ENABLE_BENCHMARKING:
https://github.com/OpenMS/OpenMS/blob/a92b5d0c6cdf931cc482283975cfd8709034411e/src/tests/CMakeLists.txt#L38

**Comment by hroest** [comment]
no these should mainly be for benchmarking but I feel that they should be tested nonetheless, e.g. if the API changes

still, I dont honestly think that benchmarking is currently the main issue in OpenMS, I would actually rather focus on integration testing and workflow testing on larger files to check whether we maintain consistent number of features and ids over time.

**Comment by jpfeuffer** [comment]
Yes, I think regularly benchmarking is a great thing. My suggestion was mainly, if the executable is only for benchmarking, then it would make sense to be able to e.g. group the benchmark tests, to easily skip them e.g. during CI and only activate them in the nightlies or something like this.

**Comment by hroest** [comment]
yes, and if we want to do real benchmarking we need bigger files where we would actually see differences, e.g. several gigabyte files and not just test files where changes in execution time will not be significant

**Comment by hroest** [comment]
PS: should we merge?

**Comment by jpfeuffer** [comment]
+1 we can still think about regular benchmarks, that are not tests, when CI is stable. For now they are just tested as simple tests

**Comment by hroest** [comment]
yes I agree. lets merge and then push forward with this...

**Pull request merged**